### ISSUE TITLE
Decrypt messages don't happen

### ISSUE THREAD

**Issue by vitormsilva** [reTHINK-project/dev-runtime-core#147]
We are developing the `resumeReporters()` and `resumeObservers()`, and when we don't execute the resume, all the encrypt and decrypt messages working will, but when we resume, the messages changed between hyperties, failling to decrypt messages from the observer side;

This prints could help you:

- Print from an observer without resume
![screenshot from 2016-12-22 10-29-03](https://cloud.githubusercontent.com/assets/10596749/21423651/1cd8a038-c836-11e6-9814-68668266f900.png)

- Print from an observer with resume
![screenshot from 2016-12-22 10-28-57](https://cloud.githubusercontent.com/assets/10596749/21423650/1cd8482c-c836-11e6-85c5-f930d6642dab.png)

as you see the `decrypted Value,` don't appears, and we have an error from side policy authorise method.

If you need extra information let me know..

**Comment by pchainho** [comment]
As we understand this is caused because keys used in the Mutual Authentication are not reused ie persisted

@tiagolb can you update us on this pls?

This one is urgent!!

**Comment by tiagolb** [comment]
I'm having a problem with using the keys correctly after retrieving these keys from the StorageManager (where they are stored for persistence).

As you can see in this picture, the error is that the key format being used is not recognized by the underlying crypto library, even though it is the same key format used before starting to store the key persistently.

![duvidagil1](https://cloud.githubusercontent.com/assets/8178030/21858054/8f0cdf5e-d81f-11e6-94da-d40dc2a0a8ae.png)

According to the documentation for the crypto library (available here: https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/importKey), the importKey function where this is failing can receive a byte array when the 'raw' specifier is used, which is what is implemented. But still I'm getting this error.

Any suggestion?

**Comment by tiagolb** [comment]
This is the function which calls the `importKey` function described in the link above:

`_importAESkey(arrayBuffer) {
    return new Promise(function(resolve, reject) {
      crypto.subtle.importKey(
          'raw', //can be 'jwk' or 'raw'
          arrayBuffer,
          {   //this is the algorithm options
            name: 'AES-CBC'
          },
          true, //whether the key is extractable (i.e. can be used in exportKey)
          ['encrypt', 'decrypt'] //can be 'encrypt', 'decrypt', 'wrapKey', or 'unwrapKey'
      )
      .then(function(key) {
        //returns the symmetric key
        //console.log('crypto-importAESkey', key);
        resolve(key);
      })
      .catch(function(err) {
        console.error('crypto-importAESkey', err);
        reject(err);
      });
    });
  }`

**Comment by tiagolb** [comment]
I have realized the the culprit of this error is the use of `JSON.parse(JSON.stringify(_this.dataObjectSessionKeys[dataObjectURL]));`. If in the original working file,  found in the repo, I add this line I get the same error.

But because I'm storing the `dataObjectSessionKey` in the StorageManager and the Sandboxes deny me to directly store and retrieve the information from the StorageManager I have used this strategy. Is there another way to store data in the StorageManager without the DataClone error causes by the Sandbox?

**Comment by tiagolb** [comment]
I have fixed the key format error and the persistence seems to be working. I have only implemented persistence for the session keys, until I make sure this is enough to ensure the issue is fixed I will keep it open.

The first impression is that this solves the issue since I can refresh the observer and the messages exchanged afterwards are correctly encrypted and decrypted.

**Comment by pchainho** [comment]
it is fixed

**Comment by vitormsilva** [comment]
I think this issue need to be reopened..

Due a fix did on the PEP, now we have again an error related with the mutual Auth.

### Steps to reproduce (common to both):
1. Open two chrome windows
2. Load the group-chat-manager  hyperty
3. Invite the other participant..

#### Resuming the observer window, until now all will working
4. refresh the window of invited participant (observer)
5. load the group-chat-manager hyperty
now you will have an error:
![screenshot from 2017-02-22 09-45-35](https://cloud.githubusercontent.com/assets/10596749/23209146/dd45313a-f8ef-11e6-9460-098417f24762.png)

#### Resuming the reporter window, steps to reproduce:
4. refresh the window of who invite (reporter)
5. load the group-chat-manager hyperty;
now you will have an error:
![screenshot from 2017-02-22 09-46-19](https://cloud.githubusercontent.com/assets/10596749/23209147/dd4528d4-f8ef-11e6-96f8-87c3f565d25e.png)

**Comment by tiagolb** [comment]
What was the fix you did to the PEP?

**Comment by vitormsilva** [comment]
To handle the forward messages - https://github.com/reTHINK-project/dev-runtime-core/blob/063da11b9776c008950f762991a3561ff5151eb1/src/policy/PEP.js#L143-L144

**Comment by pchainho** [comment]
pls give priority to this one.

**Comment by vitormsilva** [comment]
@tiagolb,

Now in the branch store-data-objects, you have the [store-object](https://github.com/reTHINK-project/dev-runtime-core/blob/store-data-objects/src/runtime/RuntimeUA.js#L141) component on the Runtime, now you can pass the `storeDataObjects` to inside the class you want.

After that, you can call the method [getDataObject](https://github.com/reTHINK-project/dev-runtime-core/blob/store-data-objects/src/store-objects/DataObjectsStorage.js#L209) and it will return an object with the info related with the dataObjectURL passed.

You can use the develop branch of dev-hyperty-toolkit.

If something is not right, please let me know, to fix it soon as possible.

**Comment by tiagolb** [comment]
@vitormsilva I've used the function you implemented to overcome the problem I was having in this issue.

But now, when the reporter sends a message to the observer this shows up:

![decrypt-error](https://cloud.githubusercontent.com/assets/8178030/24114222/5920d956-0d97-11e7-927a-7a89a7a8acec.png)

By analysing this log I think the problem arises before the message reaches the IDM. Do you have any suggestion?

**Comment by vitormsilva** [comment]
@tiagolb,

One note before, the branch `store-data-objects` don't exist anymore, it was merged into the develop branch, because we need some fixes done in this branch, sorry.
You can rename your current branch with `git branch -m <newname>`.

So, how can I test that issue, to see if I can help you?

**Comment by tiagolb** [comment]
@vitormsilva Thanks. I'll commit my changes to the repository for you to test. I'll tell you when I've done this.

**Comment by tiagolb** [comment]
I've fixed this issue [here](https://github.com/reTHINK-project/dev-runtime-core/pull/183).

Approve this pull request (the tests are failing in develop, so the tests on this branch as failing as well).

**Comment by pchainho** [comment]
We are still having problems with data objects resume when mutual authentication is used, since messages are stored encrypted and when resumed we need to decrypt them: how can we do that?

**Comment by tiagolb** [comment]
I've to take a look at how the messages are stored and maybe support a function that retrieves these messages so other components can access these.

**Comment by pchainho** [comment]
@tiagolb has provided functions at IDM to decrypt and encrypt data object and the sync manager was also updated to use it. 

From the Core Runtime everything seems to work fine and the syncher receives data objects decrypted to be resumed however, somehow, just before the DataObjectChild is instantiated, it appears again encrypted ... black magic  

![pasted image at 2017_04_20 10_42 pm](https://cloud.githubusercontent.com/assets/3893553/25290252/26924564-26c5-11e7-93c3-e6e6098501b3.png)

**Comment by pchainho** [comment]
The resume of encrypted data objects is working fine
